### ISSUE TITLE
'unicode-ipa-syls' does not preserve syllable breaks when running '--phones'

### ISSUE THREAD

**Issue by necarlson97** [ssb22/lexconvert#8]
I understand that `unicode-ipa-syls` was added for --phones2phones conversion from CMU - but would also be greatly helpful if this could be used to generate syllable separated phonemes from text, or for use with other formats.

Apologies if there is a technical limitation here - I am personally unsure of how to extract syllables from espeak phonemes, and perhaps that explains why the current behavior does not provide syllable breaks. (Assumedly CMU has some method?)

My ultimate goal is to extract syllable phonetics from whole-sentence text. (Of course, there are other tools to separate text by syllable - but generating phonemes from already-split text degrades the quality of the output greatly. Espeak seems to do a great job of identifying heteronyms and other context-specific phonetic rules when fed sentences rather than individual words/syllables. And so I am looking for a format that will preserve both the specific phonemes and the syllable breaks.)
So you can see why I was elated to find unicode-ipa-syls, and deflated to see it will not work.

(I am no linguist, forgive me if I misunderstand anything. There may be a trivial answer to this)

In any case, whether it is possible to get 'unicode-ipa-syls' to work more broadly or not, it might at least be worth a note in the docs.
And thank you again for sharing your work - it has already been greatly useful to me (and many others).

Just an example of the current behavior:
```
lexconvert --phones unicode-ipa "heretic"
hˈɛɹətˌɪk

lexconvert --phones unicode-ipa-syls "heretic"
hˈɛɹətˌɪk

lexconvert --phones2phones unicode-ipa unicode-ipa-syls "hˈɛɹətˌɪk"
hˈɛɹətˌɪk

lexconvert --phones2phones espeak unicode-ipa-syls "h'Er@t,Ik"
hˈɛɹətˌɪk

espeak -x -q "I'm excited to live in a time with live music"
 aIm Eks'aItI2d t@ l'Iv I2n a# t'aIm wID l'aIv mj'u:zIk

lexconvert --phones2phones espeak unicode-ipa-syls "aIm Eks'aItI2d t@ l'Iv I2n a# t'aIm w'Iht l'aIv mj'u:zIk"
aɪm ɛksˈaɪtɪd tə lˈɪv ɪn ə tˈaɪm wɪð lˈaɪv mjˈʉːzɪk
```

**Comment by ssb22** [comment]
Thanks, I should probably be clearer in the documentation that "syllable separators preserved" means "if we *have* them, we will keep them".  It doesn't mean we can create them if we don't have them to start with.  And, while the eSpeak format does have `%` as a syllable separator, this `%` is not usually included in the output of eSpeak's own `-X` option, which is what we use to obtain phonemes from input text, and hence the syllable separators will still be missing from `--phones`.  If you did `lexconvert --phones2phones espeak unicode-ipa-syls "mj'u:%zIk"` (note the added `%`) you should get `mjˈʉː.zɪk` (with `.` as a syllable separator), but that's because we manually added the `%` in the right place which eSpeak itself does not do.  That's why I also wrote a `--syllables` option that *tries* to split words for typesetting song lyrics using some of my own heuristics on the eSpeak output, but it's not very good: I can put up with the quality of its output in song sheets I use myself but I need to proof-read it before showing anyone else.  I'd really like to find a better solution to this if we can think of one!

**Comment by necarlson97** [comment]
Understood! Thank you for your info - it makes perfect sense. And yeah, a small note in the readme should do the trick.

I do have my own workaround for splitting syllables phonetically, and though I am a bit ashamed of it's sloppiness, I'll share it here just incase it is useful to you or others.

TL;DR:
1. Split text to syllables (I use pyphen + a fallback of [syllable corpus](https://github.com/brennyD/DeepSyllable/blob/master/Syllables.txt) as a lookup dict. If any word falls through both of those, I can always just added it to the Syllables.txt myself)
2. Merge espeak characters to individual phonemes (this was sloppily done - I was just following my nose here, I'm sure someone who understands the espeak documentation could write this better)
3. Proportionally split the phonemes the same way the text letters are split. There will be more or less phonemes than letters, but they are likely to take up roughly the same proportion. (Again, I feel someone who understands more about language could likely do a better job)
```
def get_hyphenated(word):
    # Normalize the word (lowercase and remove trailing punctuation for lookup)
    word_cleaned = re.sub(r'[.,!?]+$', '', word.lower())

    dic = pyphen.Pyphen(lang='en')
    hyphenated = dic.inserted(word_cleaned)

    # If it wasn't hyphenated by pyphen, or syllables.txt might catch it
    hyphenated = SYLLABLES_DICT.get(word_cleaned, hyphenated)

    return hyphenated
    
def get_phoneme_symbols(espeak_phonemes):
    """
    Return the espeak phonemes as a list of each separate phoneme.
    This method uses multiple passes to ensure prefixes,
    suffixes, and vowel groups are merged correctly.
    """
    # Sometimes, a word's vocalization is totally omitted:
    if espeak_phonemes == "":
        return []

    # Step 1: Split espeak_phonemes into single-character strings
    symbols = list(espeak_phonemes)

    # Step 2: Merge prefixes
    utility_prefix = "',"
    merged = []
    for char in reversed(symbols):
        if char in utility_prefix and merged:
            merged[-1] = char + merged[-1]
        else:
            merged.append(char)
    symbols = reversed(merged)

    # Step 3: Merge suffixes
    # (note: includes both the speak symbols and our
    # windows-filename-safe replacements)
    utility_suffix = "#%$=:;_|(123456789!"
    merged = []
    for char in symbols:
        if char in utility_suffix and merged:
            merged[-1] += char
        else:
            merged.append(char)
    symbols = merged

    # Step 4: Merge vowel groups (likely part of the same syllable)
    vowels = "aeiou"
    merged = []
    last_char_alpha = ""
    for char in symbols:
        # Ignore prefix/suffes when checking if it is a vowel sound
        char_alpha = "".join(
            c.lower() for c in char
            if c not in utility_prefix + utility_suffix
        )
        if merged and char_alpha in vowels and last_char_alpha in vowels:
            merged[-1] += char
        else:
            merged.append(char)
        last_char_alpha = char_alpha

    return merged


def get_phoneme_syllables(hyphenated, espeak_phonemes):
    """
    Splits the espeak_phonemes string into chunks
    matching the hyphenated syllables. I.e.:
    separate the espeak phonemes to individual sounds
    split them based on the 'size' of each syllable.
    So if the hyphenated is:
    'di-vine' (2/6 and 4/6)
    but there are 12 phoneme symbols*, it will split them
    by the proportional 4/12 and 8/12

    (*not true, just to make math easy)
    """
    split_phonemes = get_phoneme_symbols(espeak_phonemes)
    syllables = hyphenated.split('-')
    total_syllable_len = sum(len(s) for s in syllables)
    proportions = [len(s) / total_syllable_len for s in syllables]

    total_phonemes = len(split_phonemes)
    result = []
    index = 0

    for i, proportion in enumerate(proportions):
        # Calculate the approximate number of phonemes for this syllable
        if i == len(proportions) - 1:  # Last syllable gets the rest
            chunk_size = total_phonemes - index
        else:
            chunk_size = round(proportion * total_phonemes)

        # Collect the phoneme chunk
        chunk = split_phonemes[index: index + chunk_size]
        result.append(''.join(chunk))
        index += chunk_size

    return result
    ```
    
This is an imperfect solution, and not rigorously unit tested. However, it has worked so far for the (50-100) words I've tested with.
However, maybe it can be some use to you or others as a starting point.

Thank you again for your comprehensive report, and for sharing your library - it is truly appreciated!
